# Pandas [DataFrames](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) 2
- Conditional selection
- Resetting an index
- Multi-index and hiearchy


In [1]:
import numpy as np
import pandas as pd

In [2]:
from np.random import randn  #generate random numbers from normal distribution (mean = 0, variance = 1)

In [3]:
np.random.seed(101)    #just to make sure we get the same 'random' numbers

Create a DataFrame named 'df'

In [5]:
df = pd.DataFrame(data=randn(5,4),index=['A','B','C','D','E'],columns=['W','X','Y','Z'])

In [6]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Conditional selection

You can use a comparison operator against your dataframe and a dataframe of boolean values will be returned

In [9]:
bool_df = df > 0

In [10]:
bool_df

W      X      Y      Z
A   True   True   True   True
B   True  False  False   True
C  False   True   True  False
D   True  False  False   True
E   True   True   True   True

If you pass this to your original dataframe with bracket notation then you will get the values where the argument was true and null/NaN where it was false as a result

In [11]:
df[bool_df]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118       NaN       NaN  0.605965
C       NaN  0.740122  0.528813       NaN
D  0.188695       NaN       NaN  0.955057
E  0.190794  1.978757  2.605967  0.683509

If you specify a column a series of booleans will be returned

In [12]:
df['W']>0

A     True
B     True
C    False
D     True
E     True
Name: W, dtype: bool

If you pass this series into a dataframe using bracket notation it will return the *rows* where this argument is true

In [13]:
df[df['W']>0]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

As an example selecting the rows from X and Y where the values are greater than zero in column W would look like this

In [15]:
df[df['W']>0][['Y','X']]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

Breaking that command down line by line would look like the following

In [17]:
boolser = df['W'] > 0   # create a series of booleans where values are greater than zero in W
result = df[boolser]    # cast the result of the series to your DataFrame
mycols = ['Y','X']      # define a variable of series X and Y
result[mycols]          # select the columns X and Y from your DataFrame 'result'

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

You can also use multiple conditions

In [18]:
df[(df['W']>0) & (df['Y']>1)]

W         X         Y         Z
E  0.190794  1.978757  2.605967  0.683509

Notice the ampersand '&': using 'and' will return an error

In [19]:
df[(df['W']>0) and (df['Y']>1)]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

For 'or' use the pipe operator '|'

In [20]:
df[(df['W']>0) | (df['Y']>1)]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Resetting an Index

In [21]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Reset the index back to the default by calling the method df.reset_index()

In [22]:
df.reset_index()    #notice how the original index is reset to a column named 'index'

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

In [23]:
df  #notice how like other Pandas DataFrame operations you have to specify for it to occur inplace i.e. df.reset_index(inplace=True)

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

Create a new index for a DataFrame

In [24]:
new_index = ['Ayy','Bee','Cee','Dee','Eee'] #define a list of new index names

In [25]:
df['New'] = new_index   #set the list as a new DataFrame column

In [26]:
df

W         X         Y         Z  New
A  2.706850  0.628133  0.907969  0.503826  Ayy
B  0.651118 -0.319318 -0.848077  0.605965  Bee
C -2.018168  0.740122  0.528813 -0.589001  Cee
D  0.188695 -0.758872 -0.933237  0.955057  Dee
E  0.190794  1.978757  2.605967  0.683509  Eee

If you have a column in your DataFrame that you want to use as the index use the .set_index() method

In [27]:
df.set_index('New')

W         X         Y         Z
New                                        
Ayy  2.706850  0.628133  0.907969  0.503826
Bee  0.651118 -0.319318 -0.848077  0.605965
Cee -2.018168  0.740122  0.528813 -0.589001
Dee  0.188695 -0.758872 -0.933237  0.955057
Eee  0.190794  1.978757  2.605967  0.683509

Multi-Index and Hierarchy

In [28]:
# Index Levels - this is what will be used to build our Multi_Indexed DataFrame
outside = ['G1','G1','G1','G2','G2','G2']
inside = [1,2,3,1,2,3]
hier_index = list(zip(outside,inside))
hier_index = pd.MultiIndex.from_tuples(hier_index)  #method that takes in a list of tuples to create a Multi-Index

In [32]:
hier_index

MultiIndex([('G1', 1),
            ('G1', 2),
            ('G1', 3),
            ('G2', 1),
            ('G2', 2),
            ('G2', 3)],
           )

In [30]:
df = pd.DataFrame(data=randn(6,2),index=hier_index,columns=['A','B'])    #creates a 6 row by 2 column data frame using hier_index as the index and A and B as the column headings

In [31]:
df

A         B
G1 1  0.302665  1.693723
   2 -1.706086 -1.159119
   3 -0.134841  0.390528
G2 1  0.166905  0.184502
   2  0.807706  0.072960
   3  0.638787  0.329646

In [34]:
df.loc['G1']

A         B
1  0.302665  1.693723
2 -1.706086 -1.159119
3 -0.134841  0.390528

In [39]:
df.loc['G1'].loc[1]

A    0.302665
B    1.693723
Name: 1, dtype: float64

In [50]:
df.loc['G1']['A']

1    0.302665
2   -1.706086
3   -0.134841
Name: A, dtype: float64

In [41]:
df.index.names #by checking the index names we can see they are currently unnamed

FrozenList([None, None])

In [51]:
df.index.names = ['Groups','Numbers']   #set the index names

In [52]:
df

A         B
Groups Numbers                    
G1     1        0.302665  1.693723
       2       -1.706086 -1.159119
       3       -0.134841  0.390528
G2     1        0.166905  0.184502
       2        0.807706  0.072960
       3        0.638787  0.329646

As an example say we wanted to select the value 0.072960
It's at G2, Number 2, Column B
Starting from the outside and working inwards

In [54]:
df.loc['G2'].loc[2]['B']

0.07295967531703869

A useful method to know for multi-level indexing is the [cross-section](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.xs.html) method .xs()
Specify the value you are looking for and then pass in the name of the index as a second argument

In [56]:
df.xs(1,level='Numbers') #say you want to grab all values of inside index number 1 in both groups

A         B
Groups                    
G1      0.302665  1.693723
G2      0.166905  0.184502